csrc/executor_kernel_arg.cpp

```diff
@@ -334,133 +334,140 @@ void KernelArgumentHolder::pushTensorProxy(
   arguments_.push_back(getAbstractTensorArg(at::Tensor(meta_tensor)));
 }
 
-std::vector<std::byte> getTensorArgBuffer(
-    const PolymorphicValue& metadata,
+std::vector<std::byte> polymorphicValueToBytes(
+    const PolymorphicValue& argument,
+    const DataType& dtype,
     PrimDataType index_type) {
-  auto struct_ = metadata.as<Struct>();
-  std::vector<std::byte> buffer;
-  void* ptr = (void*)struct_["data"];
-  std::vector<int64_t> sizes = (std::vector<int64_t>)struct_["logical_size"];
-  std::vector<int64_t> strides = (std::vector<int64_t>)struct_["alloc_stride"];
-  if (index_type == PrimDataType::Int) {
-    buffer.reserve(
-        sizeof(ptr) + sizeof(int64_t) * (sizes.size() + strides.size()));
-    buffer.insert(
-        buffer.end(), (std::byte*)&ptr, (std::byte*)&ptr + sizeof(ptr));
-    buffer.insert(
-        buffer.end(),
-        (std::byte*)sizes.data(),
-        (std::byte*)sizes.data() + sizeof(int64_t) * sizes.size());
-    buffer.insert(
-        buffer.end(),
-        (std::byte*)strides.data(),
-        (std::byte*)strides.data() + sizeof(int64_t) * strides.size());
-  } else {
+  if (argument.is<Struct>()) {
+    auto dtype_ = std::get<StructOf>(dtype.type);
```

Comment:
It wasn't clear to me that dtype must be a `StructOf` describing the layout whenever `argument` is a `Struct`. Perhaps we should add an assert here with an informative error message for when someone like me misuses this.

Reply:
`std::get` already throw if the type mismatch. But I agree that having an explicit assert is more readable and provides better error report. Added.